// @erc7824/nitrolite/dist/rpc/types/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RPCAppStateIntent = exports.RPCProtocolVersion = exports.RPCTxType = exports.RPCChannelStatus = void 0;
var RPCChannelStatus;
(function (RPCChannelStatus) {
    RPCChannelStatus["Open"] = "open";
    RPCChannelStatus["Closed"] = "closed";
    RPCChannelStatus["Challenged"] = "challenged";
})(RPCChannelStatus || (exports.RPCChannelStatus = RPCChannelStatus = {}));
var RPCTxType;
(function (RPCTxType) {
    RPCTxType["Transfer"] = "transfer";
    RPCTxType["Deposit"] = "deposit";
    RPCTxType["Withdrawal"] = "withdrawal";
    RPCTxType["AppDeposit"] = "app_deposit";
    RPCTxType["AppWithdrawal"] = "app_withdrawal";
})(RPCTxType || (exports.RPCTxType = RPCTxType = {}));
var RPCProtocolVersion;
(function (RPCProtocolVersion) {
    RPCProtocolVersion["NitroRPC_0_2"] = "NitroRPC/0.2";
    RPCProtocolVersion["NitroRPC_0_4"] = "NitroRPC/0.4";
})(RPCProtocolVersion || (exports.RPCProtocolVersion = RPCProtocolVersion = {}));
var RPCAppStateIntent;
(function (RPCAppStateIntent) {
    RPCAppStateIntent["Operate"] = "operate";
    RPCAppStateIntent["Deposit"] = "deposit";
    RPCAppStateIntent["Withdraw"] = "withdraw";
})(RPCAppStateIntent || (exports.RPCAppStateIntent = RPCAppStateIntent = {}));
