// @erc7824/nitrolite/dist/abis/fragments.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StateParamFragment = exports.SignatureParamFragment = exports.AllocationParamFragment = exports.ChannelParamFragment = void 0;
exports.ChannelParamFragment = [
    { name: 'participants', type: 'address[2]' },
    { name: 'adjudicator', type: 'address' },
    { name: 'challenge', type: 'uint64' },
    { name: 'nonce', type: 'uint64' },
];
exports.AllocationParamFragment = [
    { name: 'destination', type: 'address' },
    { name: 'token', type: 'address' },
    { name: 'amount', type: 'uint256' },
];
exports.SignatureParamFragment = [
    { name: 'v', type: 'uint8' },
    { name: 'r', type: 'bytes32' },
    { name: 's', type: 'bytes32' },
];
exports.StateParamFragment = [
    { name: 'data', type: 'bytes' },
    {
        name: 'allocations',
        type: 'tuple[2]',
        components: exports.AllocationParamFragment,
    },
    {
        name: 'sigs',
        type: 'tuple[]',
        components: exports.SignatureParamFragment,
    },
];
